### DISCUSSION TITLE
Obtaining all available locales for a DataKey

### DISCUSSION THREAD

**Discussion by jedel1043** [unicode-org/icu4x#2374] [Q&A] [answered]
Is there a way to obtain all supported locales for a certain `DataKey`?

Expanded question:

The ECMA-402 specification requires, in all its ["Service Constructors"](https://tc39.es/ecma402/#sec-internal-slots), a way to filter a list of locales and return the subset of locales that are supported by the Service Constructor (e.g. [Intl.Collator.supportedLocalesOf](https://tc39.es/ecma402/#sec-intl.collator.supportedlocalesof)). However, to implement this algorithm, you would need the full list of supported locales for a specific SC, as shown by the `[[AvailableLocales]]` internal slot in every SC. 

Is there an API providing this functionality in ICU4X or, alternatively, plans for implementing a similar interface?

**Comment by sffc** [comment, marked as answer]
See also https://github.com/unicode-org/icu4x/issues/58 and https://github.com/unicode-org/icu4x/issues/1092

Please note that ECMA-402 never actually exposes the set of available locales, only the supportedLocalesOf function. The ICU4X data provider follows the same model; it can answer whether or not a locale is supported, but it doesn't give you the full list. It doesn't do this because there's no such thing as a list of supported locales; the set is unbounded due to locale fallback.

Therefore, please perform data requests for each locale in the list, and use that to determine whether or not to include the locale in the output list. You can either perform direct lookup (no fallback), in which case you can check for DataErrorKind::MissingLocale, or lookup with fallback, in which case you can check whether the DataResponseMetadata is "und" or not.

**Reply by jedel1043** [reply]
Thank you! That should solve it in the meantime. I will follow #58 though, in case there's an improved API for this use case in the future.

**Reply by jedel1043** [reply]
While implementing a prototype of `supportedLocalesOf`, I realized there's not a specific `DataMarker` for each ECMA-402 `Service Constructor`. Instead, each component uses a different set of `KeyedDataMarker`s to describe the requirements for that component.

Is it safe to assume all components share the same locale support for all required `KeyedDataMarker`s (which would allow me to choose one at random), or do I need to make a data request for each one of them to check if the provider fully supports that locale for that specific component?

Also, the `DateSkeletonPatternsV1Marker` returns an error if the provided locale doesn't specify the `ca` extension, even if the locale is supported by the provider. Can I assume that's intentional?

**Reply by sffc** [reply]
Indeed; this is a place where the ECMA-402 `supportedLocalesOf` simply doesn't reflect the reality of locale data, where some pieces may be available in some locales but not others.

I suggest using a certain key for each component as the reference key. I suggest:

- DateTimeFormatter: [TimeSymbolsV1Marker](https://unicode-org.github.io/icu4x-docs/doc/icu_datetime/provider/calendar/struct.TimeSymbolsV1Marker.html)
- FixedDecimalFormatter: [DecimalSymbolsV1Marker](https://unicode-org.github.io/icu4x-docs/doc/icu_decimal/provider/struct.DecimalSymbolsV1Marker.html)
- PluralRules: [CardinalV1Marker](https://unicode-org.github.io/icu4x-docs/doc/icu_plurals/provider/struct.CardinalV1Marker.html)
- ListFormatter: [AndListV1Marker](https://unicode-org.github.io/icu4x-docs/doc/icu_list/provider/struct.AndListV1Marker.html)

Continue following #58 for future work.

**Reply by sffc** [reply]
> Also, the DateSkeletonPatternsV1Marker returns an error if the provided locale doesn't specify the ca extension, even if the locale is supported by the provider. Can I assume that's intentional?

Yes, it's intentional, although this issue shouldn't come up as often since I've done some internal work on this part of the code post 1.0 Beta (see https://github.com/unicode-org/icu4x/pull/2398).

Also note that in 1.0 Final, `DateSkeletonPatternsV1Marker` will be behind an "experimental" feature. I anticipate that once it is fully stabilized, we may do like we did with the DateLengthsV1 where we make the key itself calendar-specific, rather than putting the calendar in the locale during data lookup. You can follow along in #1317.
